src/python/pants/backend/go/module.py

```diff
@@ -0,0 +1,207 @@
+    import_path: str
+    minimum_go_version: Optional[str]
+    modules: FrozenOrderedSet[ModuleDescriptor]
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class ResolveGoModuleRequest:
+    address: Address
+
+
+# Perform a minimal parsing of go.mod for the `module` and `go` directives. Full resolution of go.mod is left to
+# the go toolchain. This could also probably be replaced by a go shim to make use of:
+# https://pkg.go.dev/golang.org/x/mod/modfile
+# TODO: Add full path to expections for applicable go.mod.
+def basic_parse_go_mod(raw_text: bytes) -> Tuple[Optional[str], Optional[str]]:
+    module_path = None
+    minimum_go_version = None
+    for line in raw_text.decode("utf-8").splitlines():
+        parts = line.strip().split()
+        if len(parts) >= 2:
+            if parts[0] == "module":
+                if module_path is not None:
+                    raise ValueError("Multiple `module` directives found in go.mod file.")
+                module_path = parts[1]
+            elif parts[0] == "go":
+                if minimum_go_version is not None:
+                    raise ValueError("Multiple `go` directives found in go.mod file.")
+                minimum_go_version = parts[1]
+    return module_path, minimum_go_version
+
+
+# Parse the output of `go mod download` into a list of module descriptors.
+def parse_module_descriptors(raw_json: bytes) -> List[ModuleDescriptor]:
+    module_descriptors = []
+    for raw_module_descriptor in ijson.items(raw_json, "", multiple_values=True):
+        module_descriptor = ModuleDescriptor(
+            import_path=raw_module_descriptor["Path"],
+            module_path=raw_module_descriptor["Path"],
+            module_version=raw_module_descriptor["Version"],
+        )
+        module_descriptors.append(module_descriptor)
+    return module_descriptors
+
+
+@rule
+async def resolve_go_module(
+    request: ResolveGoModuleRequest,
+    goroot: GoLangDistribution,
+    platform: Platform,
+    bash: BashBinary,
+) -> ResolvedGoModule:
+    downloaded_goroot = await Get(
+        DownloadedExternalTool,
+        ExternalToolRequest,
+        goroot.get_request(platform),
+    )
+
+    targets = await Get(UnexpandedTargets, Addresses([request.address]))
+    if not targets:
+        raise AssertionError(f"Address `{request.address}` did not resolve to any targets.")
+    elif len(targets) > 1:
+        raise AssertionError(f"Address `{request.address}` resolved to multiple targets.")
+    target = targets[0]
+
+    sources = await Get(SourceFiles, SourceFilesRequest([target.get(GoModuleSources)]))
+    flattened_sources_snapshot = await Get(
+        Snapshot, RemovePrefix(sources.snapshot.digest, request.address.spec_path)
+    )
+    if (
+        len(flattened_sources_snapshot.files) not in (1, 2)
+        or "go.mod" not in flattened_sources_snapshot.files
+    ):
+        raise ValueError(f"Incomplete go_module sources: files={flattened_sources_snapshot.files}")
```

Comment:
This can now be deleted, thanks to Target API validation

Reply:
Deleted.